File: sentinel/src/fingerprint.ts
/**
 * Fingerprinting for privacy-preserving telemetry.
 *
 * Sentinel never ships raw user content. Instead, it ships a short hash of
 * the first 4 KB so duplicate detections can be correlated without
 * reversing the original input.
 *
 * See `docs/specs/wire-format.md` §4.
 */

import { createHash } from "node:crypto";

const FINGERPRINT_BYTES = 4096;
const FINGERPRINT_HEX_LEN = 16;

/** Compute a 16-hex-char fingerprint of the first 4 KB of `text`. */
export function fingerprint(text: string): string {
  const truncated = text.length > FINGERPRINT_BYTES
    ? text.slice(0, FINGERPRINT_BYTES)
    : text;
  return createHash("sha256")
    .update(truncated, "utf8")
    .digest("hex")
    .slice(0, FINGERPRINT_HEX_LEN);
}

/** Compute a 16-hex-char fingerprint of an arbitrary serializable object. */
export function fingerprintObject(obj: unknown): string {
  return fingerprint(JSON.stringify(obj));
}

File: sentinel-py/src/lictor_sentinel/fingerprint.py
"""
Privacy-preserving fingerprint for telemetry.

Mirrors @lictor/sentinel's src/fingerprint.ts. SHA-256 of the first 4 KB
of the input/output, truncated to 16 hex characters. See
`docs/specs/wire-format.md` §4 for the privacy invariants.
"""

from __future__ import annotations

import hashlib

FINGERPRINT_BYTES = 4096
FINGERPRINT_HEX_LEN = 16


def fingerprint(text: str) -> str:
    """16 hex characters of sha256 over the first 4 KB of `text`."""
    truncated = text[:FINGERPRINT_BYTES].encode("utf-8", errors="replace")
    return hashlib.sha256(truncated).hexdigest()[:FINGERPRINT_HEX_LEN]
